### ISSUE TITLE
multiple routes with same path

### ISSUE THREAD

**Issue by rodsouto** [zendframework/zend-expressive#40]
Hi, I tried to do this

$app->get('/users', 'Users::FetchAll');
$app->post('/users', 'Users::Create');

but the last route overwrites the first one because in https://github.com/zendframework/zend-expressive/blob/master/src/Router/Aura.php#L65 it's using the path as the route name

I'm currently using Aura.Router in other project and I give each route a different name to be able to call $router->generate($routeName, $params), I don't know if you would be interested in this, but at least if we use a combination of path+method as route name it will fix the problem

**Comment by weierophinney** [comment]
I'm thinking we have a couple paths forward.

First, I will create both `setName()` and `getName()` methods on the `Route` class.

The implementation of `getName()` will use the path + *imploded*, delimited methods, in a specific order. Thus, if no name is set on the route, it will have one that can be discovered, but also, more importantly have a unique name.

Obviously, if `setName()` is called, that name will be used.

```php
$app->get('/users', 'Users::FetchAll')->setName('fetch-users');
```

Alternately, the name can be an additional argument. The only problem with that approach, I feel, is that it makes the implementation of `route()` far more difficult and less intuitive; if you want to allow all methods, you have to pass a value for the third argument.

Next, in the `Container\ApplicationFactory`, we have two options:

- Using the route key as the name:

  ```php
  'routes' => [
      'fetch-users' => [
          'path' => '/users',
          'middleware' => 'Users::FetchAll',
      ]
  ]
  ```

- Or via a "name" key in the route itself:

  ```php
  'routes' => [
      [
          'name' => 'fetch-users',
          'path' => '/users',
          'middleware' => 'Users::FetchAll',
      ]
  ]
  ```

I'm leaning towards the second, as it's more explicit, and then we don't have to worry about determining if an integer index is intended as the name, or if we should auto-generate a name.

I'll start working on this shortly.

**Comment by ezimuel** [comment]
@weierophinney I was testing this issue and I found that also Zf2 router adapter has the same problem. I provided this test to add in RouteMiddlewareTest: https://gist.github.com/ezimuel/56ae3d169c2d3fc2b82d.

**Comment by rodsouto** [comment]
:+1: to setName() 

:+1: to "name" key in the route config

They are both the more explicit solutions :)

Thanks @weierophinney!

**Comment by weierophinney** [comment]
@ezimuel I didn't get to this today; if you want to take it while I'm on vacation, go for it; otherwise, I'll address it as soon as I return.

**Comment by ezimuel** [comment]
@weierophinney ok, I'll work on this.

**Comment by mwillbanks** [comment]
@weierophinney @ezimuel I wanted to add in my comments here as well, sometimes i have a route name and i want that to become the path if one is not set.  For instance say name is "auth" path could then become "/auth".  This is something that i already do right now.

As with child routing (which is something I had added but it's going to be far more complex to add in now so I don't have a PR yet), I used the names and / or paths to create these children.  This ended up working out surprisingly well for a multitude of cases.

**Comment by ezimuel** [comment]
I fixed this issue with PR #73. 
@mwillbanks your idea sounds interesting, we can implement it for assembling a route starting from the name, I just added the name support in `Router/Route`.

**Comment by mwillbanks** [comment]
@ezimuel perfect; yeah, I use named routes quite a bit for the assembly; when I had children I utilized an array with a children property.  However, not all routing systems support children and brings into question what the best path is for the more advanced features of certain routers.  I know that we expose it but since we normalize the response interface there needs to be some form of support in there as well.  Like I said; I've slowly been adding things such as this as I've converted things over.

**Comment by weierophinney** [comment]
Fixed with #73. Routes that are not provided names are given names that include the HTTP methods they match: `/users^GET` and `/users^POST` would be used in the examples given in the issue. If the route can handle multiple methods, HTTP method names are separated with `:`: `/users^GET:POST`. If the route can handle *any* HTTP method, the route name contains now suffix.
